### ISSUE TITLE
Chapter 8: Loading Aggregate from DB causes missing _applier in Entities

### ISSUE THREAD

**Issue by iappwebdev** [PacktPublishing/Hands-On-Domain-Driven-Design-with-.NET#26]
Hey, I think there is a problem with Entities when reloading Aggregates from DB. When Pictures have been added to a ClassifiedAd and we retrieve back the aggregate from DB, the `_applier` is missing as the objects are constructed by RavenDB. The problem is that the Apply-Method is only added in the When-Method:

```
                // picture
                case Events.PictureAddedToAClassifiedAd e:
                    picture = new Picture(Apply); // Apply added
                    ApplyToEntity(picture, e);
                    Pictures.Add(picture);
                    break;
```

When ClassifiedAd is recreated from DB, Pictures have no Apply-Method given from parent ClassifiedAd. This leeds to the error `Object reference not set to an instance of an object.` when I try to resize a picture. The error occured in Entity.cs:
--> `When(@event)` commented out because of [The sample ResizePicture event is processed twice. #2](https://github.com/PacktPublishing/Hands-On-Domain-Driven-Design-with-.NET-Core/issues/2#issue-453554732)

        protected void Apply(object @event)
        {
            // When(@event);  -> Prevent double processing
            _applier(@event); // Error occures here
        }

I created a **workaround** by extending the classes `AggregateRoot`, `ClassifiedAd`, `Entity `and `ClassifiedAdRepository`.

In AggregateRoot.cs, we force to implement a method to execute when an ClassifiedAd has been loaded from DB:
```
        public abstract void Initialize();
```

So in ClassifiedAd.cs, we need to implement this method. We go through all pictures and gives them the needed `Apply` method:
```
        // Method to handle the persistence (loading)
        public override void Initialize()
        {
            if (Pictures != null && Pictures.Any())
                Pictures.ForEach(x => x.Initialize(Apply));
        }
```

Therefore, we need following changes in Entity.cs:
```
        // BEFORE: private readonly Action<object> _applier;
        private Action<object> _applier;
        ...
        // BEFORE: protected Entity(Action<object> applier) => _applier = applier;
        protected Entity(Action<object> applier) => Initialize(applier);

        // Method to handle the persistence (loading)
        public void Initialize(Action<object> applier) => _applier = applier;
```

Finally, in ClassifiedAdRepository.cs, we need to execute the `Initialize` method when a ClassifiedAd has been loaded:
```
        public async Task<ClassifiedAd> Load(ClassifiedAdId id)
        {
            var ad = await _session.LoadAsync<ClassifiedAd>(EntityId(id));
            ad.Initialize();
            return ad;
        }
```

Please let me knwo what you think about it. Maybe there is a better solution? (apart from going with event sourcing where all events in the When method get applied).

**Comment by alexeyzimarev** [comment]
Have look in closed issues, I think it's been discussed with a resolution a couple of times.

**Comment by iappwebdev** [comment]
Hm, I already I went though all issues searched for an issue concerning this behavior but I couldn't find any,

**Comment by ghost** [comment]
Yes, I also experience this issue. I've wracked my brains for some time trying to come up with a way to protect the state changes and initialize the ad once it is loaded from the database. For example, it doesn't make sense to expose the `AggregateRoot.Apply()` method in any way, and somehow pass it to `IClassifiedAdRepository.Load()`, especially since it allows an arbitrary state change to bypass our business rules. @iappwebdev's solution is better than anything I could come up with.
